### PR TITLE
Reorganize online documentation

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
This PR reorganizes the Pyomo online documentation to follow a Diataxis-inspired model.

Very little actual documentation changes: this PR was focused on getting the structure / layout correct and is intentionally leaving documentation updates to later (smaller) PRs.  That said, there are a few significant documentation (and code) changes  (itemized below).

## Changes proposed in this PR:
- Reorganize documentation around a Diataxis-inspired format
- Enable the *automatic* generation of API (module, class, and function) documentation for the entire Pyomo codebase
- Update references to BiBTeX `alpha` format
- Update `literalinclude::` links to be relative to the documentation root directory
- Resolved numerous docstring formatting errors
- Code changes
   - Improved `ConfigDict` numpysoc generation
   - Added flags for detecting when we are testing, building documentation, and serializing
   - Fixed (by hiding) docstring formatting errors in `IntEnum`
   - Resolve tempfle cleanup issues during interpreter shutdown
   - Removed use of `import pyomo.{*}` from `__init__.py`
   - 

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Review by emma58 — APPROVED** [review]
This is really exciting! :) A few comments, but none of them should hold this up, I don't think.

File: doc/OnlineDocs/explanation/philosophy/abstract_modeling.rst [hunk 1]
```diff
@@ -0,0 +1,60 @@
+Abstract Models
+---------------
+
+.. note::
+
+   TODO: this is a copy of "Abstract vs Concrete" from Getting Started.
+   This should be expanded here.
```
[the remark this answers is not in the document]
Comment (emma58) [reply doc/OnlineDocs/explanation/philosophy/abstract_modeling.rst:7]: I guess my real question here is should we duplicate that documentation in two places?

[the remark this answers is not in the document]
Comment (jsiirola) [reply doc/OnlineDocs/explanation/philosophy/abstract_modeling.rst:7]: Yes - I did mean to leave that TODO (as it is not done yet).  Ideally, one section will talk about the *philosophy* of why we support both Abstract and Concrete modeling paradigms, and then we can (maybe) remove abstract from Getting started

File: doc/OnlineDocs/howto/debugging.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Debugging Models
+================
+
+TODO
```
[the remark this answers is not in the document]
Comment (jsiirola) [reply doc/OnlineDocs/howto/debugging.rst:4]: Yes.  We should write that documentation.  I mostly tried not to write any new documentation in this PR - just reorganize things.  In this case, this was a new section that we talked about needing to add several times, so it made sense to put in the placeholder as part of this PR.

File: doc/OnlineDocs/tutorial_examples.rst [hunk 3]
```diff
@@ -1,20 +0,0 @@
-Pyomo Tutorial Examples
-=======================
-
```
[the remark this answers is not in the document]
Comment (jsiirola) [reply doc/OnlineDocs/tutorial_examples.rst:3]: This file is being deleted.  The content was merged into the main `doc/OnlineDocs/index.rst`

File: pyomo/contrib/pyros/config.py [hunk 4]
```diff
@@ -209,10 +210,9 @@ def __call__(self, obj, from_iterable=None, allow_repeats=None):
 
     def domain_name(self):
         """Return str briefly describing domain encompassed by self."""
-        return (
-            f"{self.cdatatype.__name__}, {self.ctype.__name__}, "
-            f"or Iterable of {self.cdatatype.__name__}/{self.ctype.__name__}"
-        )
+        cdt = _domain_name(self.cdatatype)
+        ct = _domain_name(self.ctype)
+        return f"{cdt}, {ct}, or Iterable[{cdt}/{ct}"
```
Comment (shermanjasonaf) [inline pyomo/contrib/pyros/config.py:215]: Typo here: missing "`]`" at the end of the return string.

File: pyomo/core/expr/compare.py [hunk 5]
```diff
@@ -216,13 +216,14 @@ def assertExpressionsEqual(test, a, b, include_named_exprs=True, places=None):
...
-       expressions when generating the prefix notation
-
-    places: Number of decimal places required for equality of floating
-            point numbers in the expression. If None (the default), the
-            expressions must be exactly equal.
+    include_named_exprs : bool
+        If True (the default), the comparison expands all named
+        expressions when generating the prefix notation
+
+    places : float
```
Comment (blnicho) [inline pyomo/core/expr/compare.py:223]:
```suggestion
    places : int
```

File: pyomo/core/expr/compare.py [hunk 6]
```diff
@@ -265,10 +266,14 @@ def assertExpressionsStructurallyEqual(
 
     b: ExpressionBase or native type
 
-    include_named_exprs: bool
+    include_named_exprs : bool
        If True (the default), the comparison expands all named
        expressions when generating the prefix notation
 
+    places : float
```
Comment (blnicho) [inline pyomo/core/expr/compare.py:273]:
```suggestion
    places : int
```

File: pyomo/util/components.py [hunk 7]
```diff
@@ -30,8 +29,13 @@ def rename_components(model, component_list, prefix):
...
     --------
-    >>> c_list = list(model.component_objects(ctype=Var, descend_into=True))
-    >>> rename_components(model, component_list=c_list, prefix='special_')
+    >>> model = pyo.ConcreteModel()
+    >>> model.x = pyo.Var()
+    >>> model.y = pyo.Var()
+    >>> c_list = list(model.component_objects(ctype=pyo.Var, descend_into=True))
+    >>> new = rename_components(model, component_list=c_list, prefix='special_')
+    >>> str(new)
+    "ComponentMap({'special_x (key=...)': 'x', 'special_y (key=...)': 'y'})"
```
Comment (blnicho) [inline pyomo/util/components.py:38]: Shouldn't this be in a `doctest` block?

**Review by blnicho — APPROVED** [review]
I found a few minor typos but otherwise I think this can be merged

File: pyomo/util/components.py (hunk 7 above)
Comment (jsiirola) [reply pyomo/util/components.py:38]: It's not necessary.  Things that look like Python interactive sessions in the Examples are automatically picked up / tested.

**Review by mrmundt — APPROVED** [review]

**Pull request merged**